manubot/cite/citekey.py

```diff
@@ -1,82 +1,171 @@
-    "pmcid": "manubot.cite.pubmed.get_pmc_csl_item",
-    "arxiv": "manubot.cite.arxiv.get_arxiv_csl_item",
-    "isbn": "manubot.cite.isbn.get_isbn_csl_item",
-    "wikidata": "manubot.cite.wikidata.get_wikidata_csl_item",
-    "url": "manubot.cite.url.get_url_csl_item",
-}
 
-"""
-Regex to extract citation keys.
-The leading '@' is omitted from the single match group.
+@dataclasses.dataclass
+class CiteKey:
+    input_id: str
+    aliases: dict = dataclasses.field(default_factory=dict)
 
-Same rules as pandoc, except more permissive in the following ways:
+    def __post_init__(self):
+        self.check_input_id(self.input_id)
 
-1. the final character can be a slash because many URLs end in a slash.
-2. underscores are allowed in internal characters because URLs, DOIs, and
-   citation tags often contain underscores.
+    @staticmethod
+    def check_input_id(input_id):
+        if not isinstance(input_id, str):
+            raise TypeError(
+                "input_id should be type 'str' not "
+                f"{type(input_id).__name__!r}: {input_id!r}"
+            )
+        if input_id.startswith("@"):
+            f"invalid citekey input_id: {input_id!r}\nstarts with '@'"
 
-If a citekey does not match this regex, it can be substituted for a
-tag that does, as defined in citation-tags.tsv.
+    @classmethod
+    @functools.lru_cache(maxsize=None)
+    def from_input_id(cls, *args, **kwargs):
+        """Cached constructor"""
+        return cls(*args, **kwargs)
 
-https://github.com/greenelab/manubot-rootstock/issues/2#issuecomment-312153192
+    @cached_property
+    def dealiased_id(self):
+        return self.aliases.get(self.input_id, self.input_id)
 
-Prototyped at https://regex101.com/r/s3Asz3/4
-"""
-citekey_pattern = re.compile(r"(?<!\w)@([a-zA-Z0-9][\w:.#$%&\-+?<>~/]*[a-zA-Z0-9/])")
+    def _set_prefix_accession(self):
+        try:
+            prefix, accession = self.dealiased_id.split(":", 1)
+        except ValueError:
+            prefix, accession = None, None
+        self._prefix = prefix
+        self._accession = accession
+
+    @property
+    def prefix(self):
+        if not hasattr(self, "_prefix"):
+            self._set_prefix_accession()
+        return self._prefix
+
+    @property
+    def prefix_lower(self):
+        return self.prefix.lower()
+
+    @property
+    def accession(self):
+        if not hasattr(self, "_accession"):
+            self._set_prefix_accession()
+        return self._accession
+
+    @property
+    def standard_prefix(self):
+        if not hasattr(self, "_standard_prefix"):
+            self._standardize()
+        return self._standard_prefix
+
+    @property
+    def standard_accession(self):
+        if not hasattr(self, "_standard_accession"):
+            self._standardize()
+        return self._standard_accession
+
+    @cached_property
+    def handler(self):
+        from .handlers import Handler, get_handler
+
+        if self.is_handled_prefix:
```

Comment:
What happens if this condition is false?

Reply:
then `return Handler(self.prefix_lower)`  executes, which returns a generic handler